### PR TITLE
Support specified pk to bulk_create

### PR DESCRIPTION
**Pull request by lntuition**
Support specified `pk` to `bulk_create`

fixed : #387 

* [x] Support specified `pk` to `bulk_create`
* [x] Add testcase
* [x] Update Changelog

### CODE REVIEW

**Comment by lntuition** [comment]
Hmm.. I need to check why postgres failed.. Maybe it works differently with sqlite?...

**Comment by grigi** [comment]
It seems that it failed here:
```
File "/home/travis/build/tortoise/tortoise-orm/tests/test_bulk.py", line 27, in test_bulk_create_mix_specified

    self.assertEqual(all_[:1000], [{"id": id_, "name": None} for id_ in range(1000)])
```
Whilst trying to generate a diff to show a nice error message. It ran out of stack space and aborted?

So it obviously didn't do the right thing. Try testing with a smaller range (e.g. 10) to debug it with?

**Comment by grigi** [comment]
Also, PostgreSQL overrides the default `_prepare_insert_statement`: https://github.com/tortoise/tortoise-orm/blob/develop/tortoise/backends/asyncpg/executor.py#L18-L28

File: tortoise/backends/base/executor.py [hunk 1]
```diff
@@ -168,14 +168,28 @@ async def execute_insert(self, instance: "Model") -> None:
...
+                values_lists_all.append(
+                    [
+                        self.column_map[field_name](getattr(instance, field_name), instance)
+                        for field_name in self.regular_columns_all
+                    ]
+                )
+
+        if values_lists:
+            await self.db.execute_many(self.insert_query, values_lists)
+        if values_lists_all:
```
Comment (grigi) [inline tortoise/backends/base/executor.py:191]: I would do the `_all` variant first, as it uses pre-generated id's and will fail if the auto-generated id's generate a value that overlaps.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -12,6 +12,7 @@ Changelog
 0.16.11
 -------
 - fix: ``sqlite://:memory:`` in Windows thrown ``OSError: [WinError 123]``
+- Support specified ``pk`` to ``bulk_create()``
```
Comment (grigi) [inline CHANGELOG.rst:15]: This is not specific enough. Try something like:

Support ``bulk_create()`` insertion of records with overridden primary key when the primary key is DB-generated

File: tests/test_bulk.py [hunk 3]
```diff
@@ -13,6 +13,21 @@ async def test_bulk_create(self):
         inc = all_[0]["id"]
         self.assertEqual(all_, [{"id": val + inc, "name": None} for val in range(1000)])
 
+    async def test_bulk_create_with_specified(self):
+        await UniqueName.bulk_create([UniqueName(id=id_) for id_ in range(1000)])
```
Comment (grigi) [inline tests/test_bulk.py:17]: I think I have an idea re PostgreSQL. The id you generate is from 0 to 999
I don't think 0 is a safe serial number. It should start from 1 onwards?

File: tests/test_bulk.py [hunk 4]
```diff
@@ -13,6 +13,21 @@ async def test_bulk_create(self):
...
+        all_ = await UniqueName.all().values("id", "name")
+        self.assertEqual(all_, [{"id": id_, "name": None} for id_ in range(1000)])
+
+    async def test_bulk_create_mix_specified(self):
+        await UniqueName.bulk_create(
+            [UniqueName(id=id_) for id_ in range(1000)] + [UniqueName() for _ in range(1000)]
+        )
+
+        all_ = await UniqueName.all().values("id", "name")
```
Comment (grigi) [inline tests/test_bulk.py:26]: Please assert that the size of all_ is 2000
Just to ensure that ID-overlap isn't happening?

File: tortoise/backends/base/executor.py (hunk 1 above)
Comment (lntuition) [reply tortoise/backends/base/executor.py:191]: That a good point! I think it will prevent `pk` overlapping.

File: CHANGELOG.rst (hunk 2 above)
Comment (lntuition) [reply CHANGELOG.rst:15]: Oh I see, really thank you for correction. :+1:

File: tests/test_bulk.py (hunk 3 above)
Comment (lntuition) [reply tests/test_bulk.py:17]: Good. I'll Change range value and will retry it

File: tests/test_bulk.py (hunk 4 above)
Comment (lntuition) [reply tests/test_bulk.py:26]: As you said, I think Changing insertion sequence will prevent ID overlapping. But we can beat the stone bridge before you cross. I'll add it soon.

**Comment by lntuition** [comment]
@grigi :+1: CI works well. I think changing the insertion sequence is valid. But just one more question? Do we need to test more bigger case with `bulk_create`? Current size is just hundreds.

**Comment by grigi** [comment]
The tests are to validate that things are working. So doing bulk inserts with a thousand records is OK for testing. I think?

Good work, btw!

**Comment by grigi** [comment]
Oh, I see, you reduced the amount so you could see what was wrong. You're welcome to bump it back to 1000 if you so wish :-)

**Comment by lntuition** [comment]
Hmm.. rollback bulk object size make wrong with postgres :(

**Comment by grigi** [comment]
Wow, that's unexpected?

The `_prepare_insert_statement` override is to return the generated PK if required with an extra `RETURNS` statement.

I had a look at the error, and it's totally different now:
```
asyncpg.exceptions.UniqueViolationError: duplicate key value violates unique constraint "uniquename_pkey"

DETAIL:  Key (id)=(2001) already exists.
```

I also see that the error happens when inserting `values_lists` AFTER it already inserted `values_lists_all`
Which makes no sense, as the DB is supposed to be in charge of generating those values?

**Comment by grigi** [comment]
Oh, postgres has an external seril incrementer. I think if you manually specify a PK it doesn't get updated?

**Comment by lntuition** [comment]
> Oh, postgres has an external seril incrementer. I think if you manually specify a PK it doesn't get updated?

@grigi Ah.. I didn't catched it. So i'm wondering Why small object size didn't hurt external seril incrementer?... It is weird...

And Can we update external seril incrementer? and is that right solution?

**Comment by grigi** [comment]
The generated sequence is probably somwhere in the 1000-2000 range, and the specified range is lower.
We do clear the table of data, but the sequence doesn't reset. So changing it to:
```
UniqueName(id=id_) for id_ in range(10000, 11000)] + [UniqueName() for _ in range(1000)]
```
or 
```
UniqueName(id=id_) for id_ in range(1, 1001)] + [UniqueName() for _ in range(1000)]
```

A proper fix would be to emit something like this to the DB:
```sql
ALTER SEQUENCE uniquename_id_seq RESTART;
```
Or we could make it auto submit that when we truncate tables for postgresql, so that test ID's are consistent?
(Which is a lot of work, so lets not do that now).

**Comment by grigi** [comment]
I'm pretty certain the issue is NOT with the code, just the test/how we test.

**Comment by lntuition** [comment]
@grigi Thank you for your kind explanation. I understand it. I will change the range of TC :smile:

**Comment by grigi** [comment]
Oh no!
Now we have this error:
```
black --check -l 100 -t py37 tortoise/ examples/ tests/ conftest.py || (echo "Please run 'make style' to auto-fix style issues" && false)
would reformat /home/travis/build/tortoise/tortoise-orm/tests/test_bulk.py
Oh no! 💥 💔 💥
1 file would be reformatted, 178 files would be left unchanged.
Please run 'make style' to auto-fix style issues
```
:sob:

**Comment by lntuition** [comment]
Oh, I was in such a hurry ! I'll fix it quickly

**Pull request merged**